Source/GSMovieView.m

```diff
@@ -0,0 +1,1338 @@
+		  }
+	      }
+
+	    if (_statusField != nil)
+	      {
+		// Show the information...
+		NSString *syncSource = (_audioPlayer && [_audioPlayer isAudioStarted]) ? @"Audio Clock" : @"System Time";
+		_statusString = [NSString stringWithFormat: @"Rendering video frame PTS: %ld | Delay: %ld us | Sync: %@ | %@",
+					  packet.pts, delay, syncSource, _flags.playing ? @"Running" : @"Stopped"];
+
+		// Show the current decoder status...
+		[_statusField setStringValue: _statusString];
+
+		// Set the position...
+		if ([_positionField respondsToSelector: @selector(setDoubleValue:)])
+		  {
+		    NSDebugLog(@"currentPosition = %f, %ld / %ld", [self currentPosition], [self getCurrentTimestamp], [self getDuration]);
+		    [_positionField setDoubleValue: [self currentPosition]];
+		  }
+	      }
+
+	    // Show status on the command line...
+	    NSString *syncSource = (_audioPlayer && [_audioPlayer isAudioStarted]) ? @"Audio Clock" : @"System Time";
+	    NSDebugLog(@"[GSMovieView] Rendering video frame PTS: %ld | Delay: %ld us | Sync: %s\r",
+		    packet.pts, delay, [syncSource UTF8String]);
+
+	    // More tolerant timing thresholds to ensure video displays
+	    // Be extra lenient during startup (first 100 frames)
+	    static int frameCount = 0;
+	    frameCount++;
+
+	    int64_t delayThreshold = (frameCount < 100) ? 100000 : 50000; // 100ms during startup, then 50ms
+	    int64_t dropThreshold = (frameCount < 100) ? 500000 : 200000; // 500ms during startup, then 200ms
+
+	    if (delay > delayThreshold)
+	      {
+		// Only delay if we're significantly ahead
+		usleep((useconds_t)delay);
+		NSDebugLog(@"[GSMovieView] Delaying frame by %ld us (frame #%d)\r", delay, frameCount);
+	      }
+	    else if (delay < -dropThreshold)
+	      {
+		NSDebugLog(@"[GSMovieView] Dropping frame - %ld us behind (threshold: %ld us, frame #%d)\r",
+		      -delay, dropThreshold, frameCount);
+		RELEASE(dict);
+		continue; // Skip this frame to catch up
+	      }
+	    else
+	      {
+		// Normal case - display the frame without delay
+		if (packet.pts % 60 == 0) // Log every 60th frame
+		  {
+		    NSDebugLog(@"[GSMovieView] Displaying frame normally - delay: %ld us (frame #%d)\r", delay, frameCount);
+		  }
+	      }
+
+	    // Decode the packet, display it and play the sound...
+	    [self decodePacket: &packet];
+	    RELEASE(dict);
+	  }
+      }
+      RELEASE(pool);
+    }
+}
+
+- (void) renderFrame: (AVFrame *)videoFrame
+{
+  uint8_t *rgbData[1];
+  int rgbLineSize[1];
+  int width = _videoCodecCtx->width;
+  int height = _videoCodecCtx->height;
+
+  rgbLineSize[0] = width * 3;
+  rgbData[0] = (uint8_t *)malloc(height * rgbLineSize[0]);
+
+  sws_scale(_swsCtx,
+	    (const uint8_t * const *)videoFrame->data,
+	    videoFrame->linesize,
+	    0,
+	    height,
+	    rgbData,
+	    rgbLineSize);
+
+  // Create bitmap with NULL planes so it allocates its own memory
+  NSBitmapImageRep *bitmap = [[NSBitmapImageRep alloc]
```

Comment:
Here we should try to avoid generating a new NSImage and NSBitmapImageRep for each frame. Or at least avoid the `memcpy`by providing `rgbData` as the first parameter to this method call.